File: javascript/packages/highlighter/src/diff-renderer.ts
import { AnsiSink } from "./ansi-sink.js"
import { DocumentBuilder } from "./document-builder.js"
import { LineWrapper } from "./line-wrapper.js"

import type { DiffHunk } from "./diff-computer.js"
import type { Document } from "./document.js"
import type { SyntaxRenderer } from "./syntax-renderer.js"
import type { ColorScheme } from "./themes.js"

export interface DiffRenderOptions {
  contextLines?: number
  showLineNumbers?: boolean
  wrapLines?: boolean
  maxWidth?: number
  truncateLines?: boolean
  highlightInlineChanges?: boolean
  removedLineStyle?: "tint" | "dim" | "none"
  singleLineStyle?: "split" | "inline" | "auto"
  layout?: "unified" | "split"
  indent?: string
}

export class DiffRenderer {
  private syntaxRenderer: SyntaxRenderer
  private colors: ColorScheme
  private documentBuilder: DocumentBuilder

  constructor(syntaxRenderer: SyntaxRenderer, colors: ColorScheme) {
    this.syntaxRenderer = syntaxRenderer
    this.colors = colors
    this.documentBuilder = new DocumentBuilder(syntaxRenderer)
  }

  /**
   * Render the change between two sources as a syntax-highlighted diff.
   *
   * @param path - File path shown above the diff, omitted when empty
   * @param original - The source before the change
   * @param modified - The source after the change
   * @param options - Optional configuration
   * @returns The rendered diff, or an empty string when the sources are identical
   */
  render(path: string, original: string, modified: string, options: DiffRenderOptions = {}): string {
    const { contextLines = 2, highlightInlineChanges = true } = options

    const document = this.documentBuilder.buildDiff(path, original, modified, { contextLines, highlightInlineChanges })

    return this.renderDocument(document, options)
  }

  /**
   * Render hunks without the sources they came from, reconstructing each side from the line
   * contents the hunks carry.
   *
   * Use this for hunks that arrived over a wire or were parsed from a unified diff. Each
   * side is highlighted as a standalone block, so a hunk that begins part-way through a tag
   * may lex slightly differently than it would with the whole file in hand.
   *
   * @param path - File path shown above the diff, omitted when empty
   * @param hunks - The hunks to render
   * @param options - Optional configuration
   * @returns The rendered diff, or an empty string when there are no hunks
   */
  renderFromHunks(path: string, hunks: DiffHunk[], options: DiffRenderOptions = {}): string {
    const { highlightInlineChanges = true } = options

    const document = this.documentBuilder.buildDiffFromHunks(path, hunks, { highlightInlineChanges })

    return this.renderDocument(document, options)
  }

  private renderDocument(document: Document, options: DiffRenderOptions): string {
    const {
      showLineNumbers = true,
      wrapLines = true,
      maxWidth = LineWrapper.getTerminalWidth(),
      truncateLines = false,
      removedLineStyle = "tint",
      singleLineStyle = "split",
      layout = "unified",
      indent = "",
    } = options

    const sink = new AnsiSink(this.syntaxRenderer, {
      showLineNumbers,
      wrapLines,
      truncateLines,
      maxWidth,
      removedLineStyle,
      singleLineStyle,
      layout,
      indent,
    }, this.colors)

    return sink.render(document)
  }
}

File: rust/herb-highlighter/src/diff_renderer.rs
use crate::ansi_sink::{AnsiSink, AnsiSinkOptions};
use crate::diff_computer::DiffHunk;
use crate::document::Document;
use crate::document_builder::{DiffDocumentOptions, DocumentBuilder};
use crate::line_wrapper::LineWrapper;
use crate::syntax_renderer::SyntaxRenderer;
use crate::themes::ColorScheme;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum RemovedLineStyle {
  #[default]
  Tint,
  Dim,
  None,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum SingleLineStyle {
  #[default]
  Split,
  Inline,
  Auto,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum DiffLayout {
  #[default]
  Unified,
  Split,
}

#[derive(Debug, Clone)]
pub struct DiffRenderOptions {
  pub context_lines: usize,
  pub show_line_numbers: bool,
  pub wrap_lines: bool,
  pub max_width: Option<usize>,
  pub truncate_lines: bool,
  pub highlight_inline_changes: bool,
  pub removed_line_style: RemovedLineStyle,
  pub single_line_style: SingleLineStyle,
  pub layout: DiffLayout,
  pub indent: String,
}

impl Default for DiffRenderOptions {
  fn default() -> Self {
    Self {
      context_lines: 2,
      show_line_numbers: true,
      wrap_lines: true,
      max_width: None,
      truncate_lines: false,
      highlight_inline_changes: true,
      removed_line_style: RemovedLineStyle::Tint,
      single_line_style: SingleLineStyle::Split,
      layout: DiffLayout::Unified,
      indent: String::new(),
    }
  }
}

pub struct DiffRenderer<'a> {
  syntax_renderer: &'a SyntaxRenderer,
}

impl<'a> DiffRenderer<'a> {
  pub fn new(syntax_renderer: &'a SyntaxRenderer, _colors: ColorScheme) -> Self {
    Self { syntax_renderer }
  }

  pub fn render(&self, path: &str, original: &str, modified: &str, options: &DiffRenderOptions) -> String {
    let document = DocumentBuilder::new(self.syntax_renderer).build_diff(path, original, modified, &document_options(options));

    self.render_document(&document, options)
  }

  pub fn render_from_hunks(&self, path: &str, hunks: &[DiffHunk], options: &DiffRenderOptions) -> String {
    let document = DocumentBuilder::new(self.syntax_renderer).build_diff_from_hunks(path, hunks, &document_options(options));

    self.render_document(&document, options)
  }

  fn render_document(&self, document: &Document, options: &DiffRenderOptions) -> String {
    AnsiSink::new(self.syntax_renderer).render(
      document,
      &AnsiSinkOptions {
        show_line_numbers: options.show_line_numbers,
        wrap_lines: options.wrap_lines,
        truncate_lines: options.truncate_lines,
        max_width: options.max_width.unwrap_or_else(LineWrapper::get_terminal_width),
        removed_line_style: options.removed_line_style,
        single_line_style: options.single_line_style,
        layout: options.layout,
        indent: options.indent.clone(),
      },
    )
  }
}

fn document_options(options: &DiffRenderOptions) -> DiffDocumentOptions {
  DiffDocumentOptions {
    context_lines: options.context_lines,
    highlight_inline_changes: options.highlight_inline_changes,
  }
}
